tests/nn/data_parallel/test_fsdp_regnet.py

```diff
@@ -9,86 +9,216 @@
 
 """ Test FSDP with regnet-like model. """
 
+import contextlib
 import random
 import tempfile
 
 import pytest
 import torch
 import torch.multiprocessing as mp
 from torch.nn import BatchNorm2d, Conv2d, Module, SyncBatchNorm
+from torch.nn.parallel import DistributedDataParallel as DDP
 from torch.optim import SGD
 
 from fairscale.nn.data_parallel import FullyShardedDataParallel as FSDP
 from fairscale.nn.data_parallel import TrainingState, auto_wrap_bn
-from fairscale.utils.testing import dist_init, skip_if_single_gpu, teardown, torch_version
+from fairscale.utils.testing import (
+    dist_init,
+    objects_are_equal,
+    rmf,
+    skip_if_single_gpu,
+    state_dict_norm,
+    teardown,
+    torch_version,
+)
+
+
+class Model(Module):
+    def __init__(self):
+        super().__init__()
+        # TODO (Min): for now, we just test pytorch sync_bn here.
+        #             this will grow into regnet; testing apex sync_bn, etc.
+        self.conv = Conv2d(2, 2, (1, 1))
+        self.bn = BatchNorm2d(2)
+
+    def forward(self, x):
+        x = self.conv(x)
+        x = self.bn(x)
+        return x
+
+
+# We get a bit fancy here. Since the scope is `module`, this is run only
+# once no matter how many tests variations for FSDP are requested to run
+# to compare with the DDP reference. For example, a single DDP
+# reference run is needed for both flatten and non-flatten param FSDP.
+#
+# Note, this runs DDP twice with and without mixed precision and asserts
+# the resulting weights are different.
+#
+# This fixture captures and returns:
+#
+#   - model state_dict before training
+#   - model data inputs
+#   - model state_dict after training
+@pytest.fixture(scope="module")
+def ddp_ref():
+    # Get a reference model state
+    model = Model()
+    state_before = model.state_dict()
 
+    # Get reference inputs per rank.
+    world_size = 2
+    iterations = 100
+    inputs = [[]] * world_size
+    for rank in range(world_size):
+        for i in range(iterations):
+            inputs[rank].append(torch.rand(2, 2, 2, 2))
+
+    # Run DDP training twice, fp and mp.
+    for precision in ["full", "mixed"]:
```

Comment:
parameteriz-able ?

Reply:
yes, it is done below for the FSDP test. Here, we loop over both cases in DDP reference so that we compute the trained weights from the same initial weights and inputs. We ensure the two results are different with an assert below.